src/graph/graph.c

```diff
@@ -680,10 +674,288 @@ void Graph_DeleteNode(Graph *g, Node *n) {
+            array_pop(multi_edges);
+
+            /* Incase we're left with a single edge connecting src to dest
+             * revert back from array to scalar. */
+            if(array_len(multi_edges) == 1) {
+                edge_id = multi_edges[0];
+                array_free(multi_edges);
+                GrB_Matrix_setElement_UINT64(M, SET_MSB(edge_id), dest_id, src_id);
+            }
+        }
+
+        // Free and remove edges from datablock.
+        DataBlock_DeleteItem(g->edges, ENTITY_GET_ID(e));
+    }
+
+    // Delete entries.
+    for(int i = 0; i < array_len(deletions); i++) {
+        deletion = deletions[i];
+        assert(GxB_Matrix_Delete(deletion.M, deletion.row, deletion.col) == GrB_SUCCESS);
+    }
+
+    int relationCount = Graph_RelationTypeCount(g);
+    uint deletion_count = array_len(deletions);
+    for(uint i = 0; i < deletion_count; i++) {
+        deletion = deletions[i];
+        GrB_Index src = deletion.col;
+        GrB_Index dest = deletion.row;
+
+        // See if source is connected to destination with additional edges.
+        bool connected = false;
+        for(int i = 0; i < relationCount; i++) {
+            R = Graph_GetRelationMatrix(g, i);
+            info = GrB_Matrix_extractElement_BOOL(&connected, R, dest, src);
+            if(info == GrB_SUCCESS) break;
+        }
+
+        /* There are no additional edges connecting source to destination
+         * Remove edge from THE adjacency matrix. 
+         * It is OK to remove entries from the adjacency matrix, as we're 
+         * not trying to extract entries from it. */
+        if(!connected) {
+            M = Graph_GetAdjacencyMatrix(g);
+            assert(GxB_Matrix_Delete(M, dest, src) == GrB_SUCCESS);
+
+            M = _Graph_Get_Transposed_AdjacencyMatrix(g);
+            assert(GxB_Matrix_Delete(M, src, dest) == GrB_SUCCESS);
+        }
+    }
+}
+
+/* Removes both nodes and edges from graph. */
+void Graph_BulkDelete(Graph *g, Node *nodes, size_t node_count, Edge *edges, size_t edge_count, size_t *edge_deleted) {
+    assert(g);
+    
+    *edge_deleted = 0;
+
+    if(node_count) _BulkDeleteNodes(g, nodes, node_count, edge_deleted);
+
+    if(edge_count) {
+        // Filter out explicit edges which were removed by _BulkDeleteNodes.
+        if(node_count) {
+            for(int i = 0; i < edge_count; i++) {
+                Edge *e = edges+i;
+                NodeID src = Edge_GetSrcNodeID(e);
+                NodeID dest = Edge_GetDestNodeID(e);
+                
+                if(!DataBlock_GetItem(g->nodes, src) || !DataBlock_GetItem(g->nodes, dest)) {
+                    /* Edge already removed due to node removal.
+                    * Replace current edge with last edge. */
+                    edges[i] = edges[edge_count-1];
+
+                    // Update indices.
+                    i--;
+                    edge_count--;
+                }
+            }
+        }
+
```

Comment:
edge_count can be reduced to 0 in the node loop by a query like:
`MATCH (a)-[e]->() DELETE a, e`
In which case we should just return here (as the first assertion in `_BulkDeleteEdges` will fail).

Reply:
👍